Generate new types

GuillaumeGomez opened the pull request:
Fixes #80.

Waiting for https://github.com/gtk-rs/gir/pull/397 to be merged.

GuillaumeGomez commented:
Waiting for https://github.com/gtk-rs/gir/pull/398 now.

GuillaumeGomez commented:
@EPashkin: this one should be ready now.

EPashkin commented:
@GuillaumeGomez Sorry, I check this in hour.

GuillaumeGomez commented:
No hurry, take your time (and thanks!).

EPashkin commented:
`pango_font_description_from_string` has nonoptional return
`pango_font_description_set_family_static` etc. seems better ignored
Don't see way to create `AttrClass`

GuillaumeGomez commented:
It's given by a function iirc. I'll check this evening.

GuillaumeGomez commented:
Updated. I added the missing getter for `AttrClass` (you're not supposed to create it).

Also, what's wrong with `pango_font_description_set_family_static`, etc.?

EPashkin commented on src/font_description.rs:
```diff
@@ -12,3 +12,11 @@ impl FromGlibPtrNone<*const ffi::PangoFontDescription> for FontDescription {
+impl FontDescription {
+    pub fn from_string(str: &str) -> FontDescription {
+        unsafe {
+            from_glib_full(ffi::pango_font_description_from_string(str.to_glib_none().0))
```
This can be done by gir: https://github.com/gtk-rs/gdk/blob/master/Gir.toml#L176-L182

GuillaumeGomez replied:
Oh nice!

GuillaumeGomez commented:
Updated.

EPashkin commented:
By https://developer.gnome.org/pango/stable/pango-Fonts.html#pango-font-description-set-family-static
`family` need be static string. Not checked other but possibility that others has same сonstraints

GuillaumeGomez commented:
So I checked, it seems to be the only function which requires to have an argument with a static lifetime.

EPashkin commented:
IMHO others is same: return value actual only when argument live.
https://developer.gnome.org/pango/stable/pango-Fonts.html#pango-font-description-copy-static
https://developer.gnome.org/pango/stable/pango-Fonts.html#pango-font-description-merge-static

GuillaumeGomez commented:
Hum... I'll put a lifetime on the returned values as well.

EPashkin commented on src/font_description.rs:
```diff
@@ -1,158 +1,22 @@
-        unsafe { ffi::pango_font_description_free(self.0) }
-    }
-}
+}
```
Better add extra newline

GuillaumeGomez commented:
It's sad, but I had to remove the "_static" functions... Too lazy to add a `PhantomData`.

GuillaumeGomez commented:
I should maybe stop here instead of adding new types again and again...

GuillaumeGomez commented on src/context.rs:
```diff
@@ -0,0 +1,20 @@
+use glib::translate::*;
+
+impl Context {
+    pub fn get_metrics<'a, 'b, P: Into<Option<&'a FontDescription>>, Q: Into<Option<&'b Language>>>(&self, desc: P, language: Q) -> Option<FontMetrics> {
```
@EPashkin: in here too, `Q` should be `Into<Option<&mut 'b Language>>>`. :p

GuillaumeGomez commented on src/font.rs:
```diff
@@ -0,0 +1,17 @@
+use glib::translate::*;
+
+impl Font {
+    pub fn get_metrics<'a, P: Into<Option<&'a Language>>>(&self, language: P) -> Option<FontMetrics> {
```
And here as well: `Into<Option<&mut 'b Language>>>`.

EPashkin commented:
https://developer.gnome.org/pango/stable/pango-Text-Processing.html#pango-context-get-metrics
and
https://developer.gnome.org/pango/stable/pango-Fonts.html#pango-font-get-metrics
also can be handled by Gir with `const` on parameter.

GuillaumeGomez commented:
It's magical! :D

GuillaumeGomez commented:
And updated.

EPashkin commented:
https://developer.gnome.org/pango/stable/pango-Text-Processing.html#pango-context-set-language
also don't need mut. And others with `language.to_glib_none_mut()` too.
As you continue adding new objects maybe add "Pango.Fontset" too?

GuillaumeGomez commented:
I remember having an issue with FontSet but I'll try re-binding it this evening.

GuillaumeGomez commented:
Updated!

EPashkin commented:
Still `language.to_glib_none_mut()` in many places
and IMHO unneeded manual `get_metrics` for `Font` and `Context`

GuillaumeGomez commented:
Oh damn, I overwrote the automatic generation...

GuillaumeGomez commented:
No more `.to_glib_none_mut()` on `language` normally.

EPashkin commented:
👍
Thanks @GuillaumeGomez
